File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:


    // pair<pair<int,int> ,bool> helper(TreeNode* root){


    //     if(root==NULL) return {{0,0},false};


    //     int leftHeight=1+helper()

    // }



    

    int helper(TreeNode*root){

        if(root==NULL) return 0;
        int lh=helper(root->left);
        int rh=helper(root->right);
        return 1+max(lh,rh);   
    }

    bool isBalanced(TreeNode* root) {
        

        if(root==NULL) return true;


        int lh=helper(root->left);
        int rh=helper(root->right);

        bool leftBalanced=isBalanced(root->left);
        bool rightBalanced=isBalanced(root->right);


        return abs(lh-rh)<=1 && leftBalanced && rightBalanced;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // gloabl varialbe to determined if tree is balanced
    boolean balanced=true;

    public int getHeight(TreeNode root){
        if(root==null) return 0;
        int lh=getHeight(root.left);
        int rh=getHeight(root.right);
        // logic for balanced
        if(Math.abs(lh-rh) >1) balanced =false;
        // standard to get height of tree (height of deepest subtree)
        return 1+Math.max(lh,rh);
    }
    public boolean isBalanced(TreeNode root) {
        getHeight(root);
        return balanced;    
    }
}